Add `edges` vector to CodeInstance/CodeInfo to keep backedges as edges

vtjnash opened the pull request:
Appears to add about ~~1MB~~ ~~3MB (117MB to 120MB)~~ 11MB (128MB to 139MB) to the system image, and to decrease the stdlib size by ~2MB~ ~54 MB (305MB to 251MB)~ 55 MB (325MB to 270MB), so seems overall favorable right now. The edges are ~not~ being computed ~yet~ following the encoding https://hackmd.io/sjPig55kS4a5XNWC6HmKSg?both#Edges-Encoding to correctly reflect the backedges ~but they currently are just extracting the `:invoke` fields~.

~~TODO: we currently store 2 copies of the edges during serialization (in both the svec and the old format) which should help recover a portion of the overhead for the stdlibs.~~

~~TODO: this records edges only to the MI, not to the recursive edges (as required for correctness)~~

vtjnash commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/54894#issuecomment-2222689945) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/8fafb91_vs_b491bcc/report.md).

oscardssmith commented:
Given that this is now a pure regression size-wise, can the description be updated to more clearly explain why this PR is a good thing?

vtjnash commented:
As for the segfault, it looks like codegen is failing to emit a write-multi-barrier in
```
julia> code_llvm(push!, (LockFreeSortedLinkedList{Vector{Any}}, Vector{Any}), optimize=false)
```
which just has an empty block
```
xchg_wb:                                          ; preds = %done_xchg
    br label %done_xchg_wb
```
resulting in a gc of the values in the linked list before they are used

aviatesk commented:
I'm working on rebasing this.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/54894#issuecomment-2373802398) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/0658390_vs_25cbe00/report.md).

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/54894#issuecomment-2374012628) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/ebe9c40_vs_6e5e87b/report.md).

aviatesk commented:
Now we have these remaining TODOs for the Julia-side refactoring:
- Finalize the format for `sv.edges`. There might be cases where no
  `edge::CodeInstance` exists as a result of `abstract_call_method`,
  and in such cases, we might still need to use `MethodInstance` in the
  `edges` list.
- Ensure that when the local caching mode is specified (i.e. for
  const-prop'ed calls and call-site-inlined calls), the const-propped
  edge should be propagated instead of the regular edge.
- Make use of the `CodeInstance` held by `CallInfo` during inlining
  for slightly better performance by avoiding the global cache lookup.

aviatesk commented on base/compiler/typeinfer.jl:
```diff
@@ -129,14 +130,22 @@ function finish!(interp::AbstractInterpreter, caller::InferenceState;
+        if istoplevel
+            edges = empty_edges
+        else
+            # TODO remove the next line once we make staticdata_utils.c able to handle `CodeInstance` edge directly
+            edges = Any[edge isa CodeInstance ? edge.def : edge for edge in caller.edges]
```
I think the Julia-side refactoring is complete for now. `edges` is now a list of `CodeInstance` instead of `MethodInstance`. By removing this one line, the C-side implementation will be able to handle recursive edges as a list of `CodeInstance`.
@vtjnash I’d like to request your review and assistance with the C-side refactor at this point, if possible.

---

Below are some concerns and ideas I have so far:

- Currently, we’re generating dummy `CodeInstance` objects with `codeinst_as_edge(...)` to represent const-propped callsite edges. These `dummy_edge::CodeInstance` objects don’t contain much useful information outside of `dummy_edge.edges`, so most of their fields are wasted. So `CodeInstance` might be a bit too heavy as an efficient encoding of the recursive edges, and we might want to consider defining a dedicated object just for recursive edges.

- Maybe related to the points above, the roles of `CodeInstance` and `InferenceResult` have become quite similar. A refactor to merge `InferenceResult` into `CodeInstance` might also be worth considering.

- In places like `_hasmethod_tfunc` and `abstract_applicable`, where backedges are needed for method match, or for inlined call sites without a proper `edge::CodeInstance`, we’re generating dummy `dummy_invoke_edge::CodeInstance`s with `codeinst_as_invoke_edge()`. They are sufficient for representing backedges, but I’m not entirely confident they're appropriate as forward edges (`dummy_invoke_edge.edges` are always empty)

- As a minor improvement for the inlining algorithm, we could potentially improve it to skip `code_cache` lookups by using the proper `edge::CodeInstance` in `CallInfo`, but I haven’t implemented that yet (or in a larger context, this will probably be done anyway as part of addressing JuliaLang/julia#54899).

topolarity replied:
> we’re generating dummy CodeInstance objects with codeinst_as_edge(...) to represent const-propped callsite edges. 

That's pretty similar to our "const return ABI" code instances, right? IIRC, those are also mostly used just for their return type + backedges.

> CodeInstance might be a bit too heavy as an efficient encoding of the recursive edges

Do we not have (the equivalent of) MethodTable edges in the new scheme? Are those just CodeInstances w.r.t. to a wide MethodInstance now?

aviatesk replied:
> That's pretty similar to our "const return ABI" code instances, right? IIRC, those are also mostly used just for their return type + backedges.

Yes, that's correct. Those `CodeInstance`s aren’t stored in the cache, and only a few of their fields are actually used.

> Do we not have (the equivalent of) MethodTable edges in the new scheme? Are those just CodeInstances w.r.t. to a wide MethodInstance now?
> 
> **edit:** I suppose abstract calls in general are covered by the "N CodeInstances" encoding of an edge but that sounds potentially inefficient?

What I was referring to is the general design of recursive edges. Currently, a `caller::CodeInstance` has `caller.edges::SimpleVector`, and `caller.edges` consists of `mt::MethodTable` /w `sig` edges or other general `edge::CodeInstance`s. However, some of these `edge::CodeInstance`s are created by `codeinst_as_edge(...)` or `codeinst_as_invoke_edge(...)` as mentioned before, and these edges contain unnecessary fields—this was the point I was trying to make. 

Still, I’m not sure if there’s a more efficient design than the current implementation.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

aviatesk commented on base/compiler/tfuncs.jl:
```diff
@@ -3010,10 +3010,6 @@ function abstract_applicable(interp::AbstractInterpreter, argtypes::Vector{Any},
-            for i = 1:napplicable
-                (; match, edges, edge_idx) = applicable[i]
-                edges[edge_idx] = codeinst_as_invoke_edge(interp, specialize_method(match))
-            end
```
`codeinst_as_invoke_edge` here was necessary to recode backedges for this method match.

vtjnash replied:
I changed the hackmd reference to reflect this change now: no specialization edges (CodeInstance) are required if the only point is to record the existence of the edge (the MethodInstance). The edge computation code was updated to be aware of this point as well, such that it will generate the method match edge as needed, but will then replace that with a CodeInstance edge if that results in redundancies.

aviatesk replied:
Yes, I think the `Method`-edge encoding in particular seems useful, where we can distinguish between edges that don’t need to be recorded as backedges when `abstract_call_method` fails and those that do need to be recorded as backedges, such as in `abstract_applicable`.

aviatesk replied:
I implemented the `Method`-edge encoding: 3836a39cc1f5c8e2f0939113a2fc0dad4e69f6f8

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

aviatesk commented on base/compiler/typeinfer.jl:
```diff
@@ -95,26 +95,36 @@ const __measure_typeinf__ = RefValue{Bool}(false)
+        result.src = ir_to_codeinf!(opt)
+    end
+    #@assert last(result.valid_worlds) <= get_world_counter() || isempty(caller.edges)
+    if isdefined(result, :ci_for_cache) || isdefined(result, :ci_as_edge)
```
`isdefined(result, :ci_as_edge)` is actually dead, since we set the field after `typeinf` in the both places where we did it.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/54894#issuecomment-2449280393) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/936b0e2_vs_2a24b8f/report.md).

vtjnash committed:
add edges metadata field to CodeInfo/CodeInstance, generate backedges from it

This records all edge metadata in a field of any CI objects so that
staticdata can directly validate those objects from just that data.

Co-authored-by: Shuhei Kadowaki <aviatesk@gmail.com>

aviatesk commented:
CI is all green. Let's merge this and move things forward.

topolarity commented:
Very excited about this! Really great work @vtjnash and @aviatesk 👍

JeffBezanson commented:
Was the inference performance regression resolved?

vtjnash commented:
There was no actual regression ever. That was simply measurement error due to something about inference being unreliable and does not reproducibly. Confirmed here:
https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_date/2024-11/01/report.md

vchuravy commented on test/precompile_absint2.jl:
```diff
@@ -64,29 +63,33 @@ precompile_test_harness() do load_path
             @test ci.max_world == typemax(UInt)
             @test Base.module_build_id(TestAbsIntPrecompile2) ==
                 Base.object_build_id(ci)
+            end
```
*sigh* skipping test is quite annoying. This test was modelling how GPUCompiler & co interact with the cache system and that code instances owned by an external compiler do actually survive precompilation.
